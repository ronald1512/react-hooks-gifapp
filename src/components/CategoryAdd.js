import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const CategoryAdd = ({setcategories}) => { //aqui usamos desestructuracion porque en realidad esto viene dentro del objeto props
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e)=>{
        setInputValue(e.target.value);
        console.log('handleInputChange llamado');
    }

    const handleSubmit =(e) =>{
        e.preventDefault(); /**Esto es para prevenir el comportamiento por defecto del formulario, que es recargar la página */

        console.log('handleSubmit');
        if(inputValue.trim().length>2){
           setcategories( cats => [inputValue,...cats]);
           setInputValue('');
        }
    }

    return (        
        <form onSubmit={handleSubmit}> {/**Este onSubmit se puede disparar cuando le dan a un boton o cuando da enter en el input */}
            <p> { inputValue }</p> {/**Esto lo agregué nomás para hacer el test de este componente */}
            <input 
                    type="text"
                    value={inputValue}
                    onChange= {handleInputChange}
                />
        </form>            
    )
}



CategoryAdd.propTypes = {
    setcategories:PropTypes.func.isRequired
}
