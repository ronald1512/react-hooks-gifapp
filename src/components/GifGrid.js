// import React, { useState, useEffect } from 'react'
import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';
// import { getGifs } from '../helpers/getGifs';


/**useEffect: permite ejecutar cierto código de forma condicional. Aqui lo vamos a usar para que
 * no ejecute el de getGifs cada vez que identifique un cambio en algun state y tenga de renderizar de nuevo. 
*/

const GifGrid = ({category}) => {
    // const [images, setImages] = useState([])
    // useEffect(()=>{
    //     getGifs(category).then(setImages);  //esto es lo mismo que haber hecho: imgs=> setImages(imgs)            -> es exactamente lo mismo
    // }, [category]); //significa que esto quiero que se ejecute solo cuando es componente es renderizado por primera vez
    
    /**
     * React Hook useEffect has a missing dependency: 'category'. 
     * Either include it or remove the dependency array  react-hooks/exhaustive-deps
     * 
     * 
     * Este warning nos indica que si la 'category' cambia, entonces tendría que volver a ejecutar lo que está dentro del useEffect. Para 
     * permitirlo, solo basta añadir category en el [] de dependencias de useEffect.
     * 
     */


    /**             Comenté lo de arriba para a continuacion, usar mi custom hook
     * 
     */

    const {data:images, loading} = useFetchGifs(category);
    



    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{ category }</h3>
            {/** usar && es una forma corta de hacer lo de abajo, si loading=true, va a evaluar el <p></p>  */}
            {loading && <p className="animate__animated animate__flash">Loading</p>}
             <div className="card-grid">
                

                {
                    images.map(img =>(
                        <GifGridItem 
                            key={img.id}
                            {...img}//de esta forma estoy mandando cada una de las propiedades de img como una propiedad independiente!!
                        />
                    ))
                }

            </div>
        </>
    )
}

//RECORDATORIO: Cuando no se ponen las llaves en una funcion de flecha, se hace un return implícito. 

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

export default GifGrid
