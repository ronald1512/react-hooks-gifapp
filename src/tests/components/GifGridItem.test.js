import { shallow } from "enzyme"
import { GifGridItem } from "../../components/GifGridItem"

describe('Pruebas en <GifGridItem />', () => {
    const title = 'Un titulo';
    const url = 'https://aws.s3/imagen.jpg';
    const wrapper = shallow(<GifGridItem title={title} url={url} />);

    test('debe de mostrar el componente correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de terner un parrafo con el title', () => {
        const p = wrapper.find('p');
        expect(p.text().trim()).toBe(title);
    });

    test('debe de tener la imagen igual al url y alt de los props', () => {
        const img = wrapper.find('img');
        //console.log(img.html());
        //console.log(img.props().src);   // para jalar el objeto con las propiedades
        expect(img.props().src).toBe(url);
        expect(img.props().alt).toBe(title);
    });

    test('debe de tener animate__fadeIn', () => {
        const div = wrapper.find('div');
        const className=div.prop('className');
        expect(className.includes('animate__fadeIn')).toBe(true);
    })
    
    
    
    
})
