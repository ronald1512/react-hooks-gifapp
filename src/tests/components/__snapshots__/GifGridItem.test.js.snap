// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem /> debe de mostrar el componente correctamente 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="Un titulo"
    src="https://aws.s3/imagen.jpg"
  />
  <p>
    Un titulo
  </p>
</div>
`;
